### PR TITLE
JIT: account for newly unreachable blocks in morph

### PR DESCRIPTION
**Pull request by AndyAyersMS**
Morph can alter control flow, if (for instance) it can prove a block's conditional branch must go a certain way.

Take advantage of this to improve morph's cross-block assertion prop, by not considering unreachable predecessors when computing the incoming assertion state. This is similar to something we already do in value numbering, but there control flow isn't being altered.

Also, when we can prove that a block has become unreachable, remove the block IR and alter its jump kind, so we are not spending time morphing IR we'll subsequently just throw away.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
We might want something like this in local morph too... see eg [this comment in #104250](https://github.com/dotnet/runtime/issues/104250#issuecomment-2202834723).

File: src/coreclr/jit/morph.cpp [hunk 1]
```diff
@@ -13430,6 +13492,11 @@ PhaseStatus Compiler::fgMorphBlocks()
         INDEBUG(fgSafeBasicBlockCreation = false;);
         INDEBUG(fgSafeFlowEdgeCreation = false;);
 
+        // We will track which blocks become unreachable during morph
+        //
+        EnsureBasicBlockEpoch();
+        BlockSet unreachable = BlockSetOps::MakeEmpty(this);
```
Comment (jakobbotsch) [inline src/coreclr/jit/morph.cpp:13498]: This can use post-order number based bit vectors.

Comment (AndyAyersMS) [reply src/coreclr/jit/morph.cpp:13498]: Good point. 

I was actually thinking I could possibly just rely on edge removal, though there are some odd cases where we artificially bump up ref counts to keep blocks from being considered unreachable.

SPMI results are way more variable across ABI/ISA than I'd expect. Not sure if this is missing collections or something more fundamental...

**Comment by AndyAyersMS** [comment]
Still vexed by BBJ_CALLFINALLY issues.

**Comment by AndyAyersMS** [comment]
[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=858093&view=ms.vss-build-web.run-extensions-tab) -- both size and TP wins.

**Comment by AndyAyersMS** [comment]
@jakobbotsch PTAL
cc @dotnet/jit-contrib 

[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=858093&view=ms.vss-build-web.run-extensions-tab) -- good size and TP wins.

Haven't looked at regressions extensively, but saw one case where morph copy-propped a normal local into uses of a "is never negative" local and so we lost the "is never negative" aspect, and this caused us to miss an RBO because of signed/unsigned differences in VN.

Will fire up some stress...

**Review by jakobbotsch — APPROVED** [review]

**Pull request merged**